### PR TITLE
Improve testing coverage

### PR DESCRIPTION
**Pull request by alexturkin**
- [x] use `mock request` for RestClient tests

### Problem Description
Our testing coverage is quite low for the library.

### Solution Description
* Moar tests! Test coverage is increased:

---
before:
```
 <class name="al_servicec.js" filename="al_servicec.js" line-rate="0.5138" branch-rate="0.425">
 <class name="al_util.js" filename="al_util.js" line-rate="0.36840000000000006" branch-rate="0">
```

after:
```
<class name="al_servicec.js" filename="al_servicec.js" line-rate="0.8273" branch-rate="0.6842">
<class name="al_util.js" filename="al_util.js" line-rate="0.6315999999999999" branch-rate="0.5">
```
---

Tests added:
* AimsC
* EndpointsC
* Al_utils

Tests for `RestServiceClient` are to follow

### CODE REVIEW

File: test/aimsc_test.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        afterEach(function() {
+            fakeRest.restore();
+            tk.reset();
+        });
+
+        after(function() {
+            clean_file_cache(m_alMock.CACHE_FILENAME);
+        });
+
+        it('authenticate with cold cache', function(done) {
```
Comment (alexturkin) [inline test/aimsc_test.js]: I am also going to add test about handling "bad credentials" case (when creds are rejected)

File: al_servicec.js [hunk 2]
```diff
@@ -49,6 +49,7 @@ class AimsC extends m_alUtil.RestServiceClient {
     }
 
     _makeAuthRequest() {
+        // console.log("!!!!!!!!!!!", this._isTokenMemCached());
```
Comment (kkuzmin) [inline al_servicec.js]: please remove.

File: al_servicec.js [hunk 3]
```diff
@@ -65,22 +66,17 @@ class AimsC extends m_alUtil.RestServiceClient {
     }
 
     _isTokenExpired(aimsToken) {
-        return aimsToken.authentication.token_expiration > (Date.now()/1000 + 600);
+        return aimsToken.authentication.token_expiration <= (Date.now()/1000 + 600);
```
Comment (kkuzmin) [inline al_servicec.js]: Could you please add `const` var for `600`

File: al_servicec.js [hunk 4]
```diff
@@ -65,22 +66,17 @@ class AimsC extends m_alUtil.RestServiceClient {
...
     }
 
     _isTokenMemCached() {
-        return this._aimsResponse && this._isTokenExpired(this._aimsResponse);
+        return (this._aimsResponse != undefined) && !this._isTokenExpired(this._aimsResponse);
```
Comment (kkuzmin) [inline al_servicec.js]: Please keep
```
return this._aimsResponse && !this._isTokenExpired(this._aimsResponse);
```
to cover _aimsResponse = null.

File: al_servicec.js [hunk 5]
```diff
@@ -65,22 +66,17 @@ class AimsC extends m_alUtil.RestServiceClient {
...
     _isTokenFileCached() {
         var filename = this._tokenCacheFile;
-        fs.readFile(filename, (readError, fileContent) => {
-            if (readError) {
-                if ( readError.code !== 'ENOENT' ) {
-                    fs.unlinkSync(filename);
-                }
-                return false;
-            }
+        try {
```
Comment (kkuzmin) [inline al_servicec.js]: Please double-check if we could have one `try..catch` here and reason about exceptions type: read or JSON parse.

File: index.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+ * -----------------------------------------------------------------------------
+ */
+
+module.exports = {
+    AimsC : require('./al_servicec').AimsC,
+    AlServiceC : require('./al_servicec').AlServiceC,
+    IngestC : require('./al_servicec').IngestC,
+    AzcollectC : require('./al_servicec').AzcollectC,
+    EndpointsC : require('./al_servicec').EndpointsC,
+    RestServiceClient : require('./al_util').RestServiceClient
```
Comment (kkuzmin) [inline index.js]: Let's not expose it for now. I don't see any use case for it.

File: test/aimsc_test.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+
+function assert_cache(aimsC, expectedMem, expectedFile) {
+    assert.equal(aimsC._isTokenMemCached(), expectedMem);
+    assert.equal(aimsC._isTokenFileCached(), expectedFile);
+}
+
+
+function malform_cache_file(filename) {
+    fs.writeFileSync(filename, "not-a-json");
```
Comment (kkuzmin) [inline test/aimsc_test.js]: please use single quotes.

File: test/aimsc_test.js [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+            tk.freeze(new Date(INITIAL_TS * 1000));
+
+            fakeRest = sinon.stub(RestServiceClient.prototype, 'post').callsFake(
+                function fakeFn(path, options) {
+                    if (path == '/aims/v1/authenticate' &&
+                        options.auth.user == m_alMock.AIMS_AUTH.auth.user &&
+                        options.auth.password == m_alMock.AIMS_AUTH.auth.password) {
+                        return new Promise(function(resolve, reject) {
+                            resolve({
+                                authentication : {
```
Comment (kkuzmin) [inline test/aimsc_test.js]: I suggest to move tocken json to `al_mock.js`

File: test/endpointc_test.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
+const assert = require('assert');
```
Comment (kkuzmin) [inline test/endpointc_test.js]: Filename `endpointSc`

File: test/al_util_test.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+const RestServiceClient = require('../al_util').RestServiceClient;
+const m_alMock = require('./al_mock');
+const debug = require('debug') ('azcollectc_test');
+var servicecRewire = rewire('../al_servicec');
+var m_servicec = require('../al_servicec');
+
+describe('Unit Tests', function() {
+
+    describe('RestServiceClient', function() {
+        it('check _initRequestOptions', function(done) {
```
Comment (alexturkin) [inline test/al_util_test.js]: @kkuzmin 
Ideally, I'd like to test every method of `RestServiceClient` here, but for that we need to make `rp`. Do you have any ideas on how to do it correctly?

I tried this:

```js

const req = require('request');


var fakeRp;

beforeEach(function() {
    fakeRp = sinon.stub(req, 'request').callsFake(
        function fakeFn(newOptions) {
            return new Promise(function(resolve, reject) {
                resolve({response : 'some_response'});
            });
        }
    );
});

afterEach(function() {
    fakeRp.restore();
});


it('check request', function(done) {
    var restC = new RestServiceClient(m_alMock.AL_API);
    var body = {
        field1 : true,
        field2: 'false'
    };
    restC.request('POST', '/some/path', { body : body })
    .then(resp => {
        sinon.assert.calledWith(fakeRp,
            {
                method: 'POST',
                url: 'https://al-api-endpoint.alertlogic.com/some/path',
                json: true,
                headers: {
                    'Accept': 'application/json'
                },
                pool: {
                    maxSockets: 128
                },
                body : body
            }
        );
        done();
    });
});
```
but it fails 😞 

```
(node:28575) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): RequestError: Error: getaddrinfo ENOTFOUND al-api-endpoint.alertlogic.com al-api-endpoint.alertlogic.com:443
(node:28575) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

File: test/al_util_test.js [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+                headers : {'some_header' : 'some_value'},
+                body : TEST_BODY
+            })
+            .then(resp => {
+                assert.ok(restMock.isDone());
+                done();
+            });
+        });
+
+        it('test post', function(done) {
```
Comment (alexturkin) [inline test/al_util_test.js]: @kkuzmin pls take a look if you're +1 to new RestClient tests

**Pull request merged**